### ISSUE TITLE
File permission error on latest build

### ISSUE THREAD

**Issue by blootsvoets** [bitnami/bitnami-docker-postgresql#192]
**Description**

Latest build of PostgreSQL contains a permission mismatch or missing file at `/opt/bitnami/postgresql/conf/postgresql.conf`. This issue is not present on earlier builds.

**Steps to reproduce the issue:**

Run the following image:
```
bitnami/postgresql          12.1.0            450022f1da59        28 hours ago        246MB
```

**Describe the results you received:**

Logs:
```
postgresql 11:29:19.95 
postgresql 11:29:19.95 Welcome to the Bitnami postgresql container
postgresql 11:29:19.95 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 11:29:19.95 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 11:29:19.96 Send us your feedback at containers@bitnami.com
postgresql 11:29:19.96 
postgresql 11:29:19.96 INFO  ==> ** Starting PostgreSQL setup **
postgresql 11:29:19.99 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql 11:29:20.00 INFO  ==> Loading custom pre-init scripts...
postgresql 11:29:20.00 INFO  ==> Initializing PostgreSQL database...
postgresql 11:29:20.00 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql 11:29:20.04 INFO  ==> pg_hba.conf file not detected. Generating it...
postgresql 11:29:20.05 INFO  ==> Generating local authentication configuration
postgresql 11:29:20.07 INFO  ==> Deploying PostgreSQL with persisted data...
postgresql 11:29:20.08 INFO  ==> Configuring replication parameters
sed: can't read /opt/bitnami/postgresql/conf/postgresql.conf: Permission denied
postgresql 11:29:20.08 INFO  ==> Stopping PostgreSQL...
```

**Describe the results you expected:**

Run without permission errors.

**Additional information you deem important (e.g. issue happens only occasionally):**

Only on the latest build. No issue on
```
bitnami/postgresql                                           12.1.0-debian-9-r58   68c7ca35754e        3 days ago          227MB
```
but the issue also occurs on
```
bitnami/postgresql                                           11.6.0                cb359f513f28        21 hours ago        245MB
```

**Version**

- Output of `docker version`:

```
lient: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea838
 Built:             Wed Nov 13 07:50:12 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea838
  Built:            Wed Nov 13 07:48:43 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

- Output of `docker info`:

```
Client:
 Debug Mode: false

Server:
 Containers: 4
  Running: 3
  Paused: 0
  Stopped: 1
 Images: 45
 Server Version: 19.03.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.4.0-166-generic
 Operating System: Ubuntu 16.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.796GiB
 Name: ucb-atlas-dev.thehyve.net
 ID: A3RI:F2FH:FNJ2:64UL:AFEN:KPIL:CYK7:A5TN:YMAH:ISZN:3W2Z:5HMD
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

Running on an EC2 instance.

**Comment by andresbono** [comment]
Hi @blootsvoets, thanks for reporting it. It seems the same issue than https://github.com/helm/charts/issues/20369. We are not able to reproduce it. We tried both versions you are reporting:

- `bitnami/postgresql:12.1.0` (Digest: `sha256:382e5ebfc04de6871a1acd7204f2b8768a65291891c27dae1848e19c6a25686f`)
- `bitnami/postgresql:11.6.0` (Digest: `sha256:37fccbd36b913f6f5b7d65dcb8f397d631bdf4890300c7dde93ff81c4eec817c`)

- Can you tell us the specific command you are running to execute the container?
- Are you mounting your own configuration file?
- Are you able to check the permissions of `/opt/bitnami/postgresql/conf/postgresql.conf` inside the container? For us:

```
$ ls -la /opt/bitnami/postgresql/conf/postgresql.conf
-rw-r--r-- 1 1001 root 26348 Jan 27 13:22 /opt/bitnami/postgresql/conf/postgresql.conf
```

With that information, I hope we get some hints so we can continue investigating it.

**Comment by blootsvoets** [comment]
We aren't mounting the postgres file. I can't exec into a container, because it exits, but I can run
```
sudo docker-compose run --rm postgresql ls -la /opt/bitnami/postgresql/conf/postgresql.conf

postgresql 15:01:02.20 
postgresql 15:01:02.20 Welcome to the Bitnami postgresql container
postgresql 15:01:02.20 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 15:01:02.20 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 15:01:02.20 Send us your feedback at containers@bitnami.com
postgresql 15:01:02.21 

ls: cannot access '/opt/bitnami/postgresql/conf/postgresql.conf': No such file or directory
```

sudo docker-compose run --rm postgresql ls -la /opt/bitnami/postgresql/conf/
postgresql 15:02:26.37 
postgresql 15:02:26.37 Welcome to the Bitnami postgresql container
postgresql 15:02:26.37 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 15:02:26.37 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 15:02:26.37 Send us your feedback at containers@bitnami.com
postgresql 15:02:26.38 
```

And the parent directory:
```
sudo docker-compose run --rm postgresql ls -la /opt/bitnami/postgresql/conf/
total 12
drwxrwxr-x 3 0 postgres 4096 Jan 26 07:59 .
drwxr-xr-x 1 0 postgres 4096 Jan 26 07:59 ..
drwxrwxr-x 2 0 postgres 4096 Jan 26 07:59 conf.d
```

My docker-compose file contains the following:
```yaml
  postgresql:
    image: bitnami/postgresql:12.1.0
#    image: bitnami/postgresql:12.1.0-debian-9-r58
    shm_size: 256m
    networks:
      - default
      - ohdsi_postgres
    ports:
      - "5432:5432"
    volumes:
      - postgresql_data:/bitnami/postgresql
      - ./lib/postgresql:/docker-entrypoint-initdb.d:ro
      - ./data/omop_vocabulary:/data/vocabulary
    environment:
      - POSTGRESQL_USERNAME=postgres
      - POSTGRESQL_PASSWORD=${POSTGRESQL_PASSWORD}
      - PGPASSWORD=${POSTGRESQL_PASSWORD}
      - POSTGRESQL_DATABASE=ohdsi
```

**Comment by blootsvoets** [comment]
Thinking it through, running without a container does not actually create the correct files. If I run the commands on the stopped container, I do get a valid file list, I'm unsure why the sed command would fail then...
```
$ sudo docker-compose up -d postgresql
$ sudo docker ps -a
CONTAINER ID        IMAGE                                                        COMMAND                  CREATED             STATUS                     PORTS                  NAMES
242aa600a062        bitnami/postgresql:12.1.0                                    "/entrypoint.sh /run…"   7 seconds ago       Exited (2) 4 seconds ago                          docker_postgresql_1
$ sudo docker commit 242aa600a062 postgres-bad
$ sudo docker run --rm postgres-bad ls -al /opt/bitnami/postgresql/conf/
postgresql 15:15:31.36 
postgresql 15:15:31.37 Welcome to the Bitnami postgresql container
postgresql 15:15:31.37 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 15:15:31.37 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 15:15:31.37 Send us your feedback at containers@bitnami.com
postgresql 15:15:31.37 

total 44
drwxrwxr-x 1        0 postgres  4096 Jan 27 15:14 .
drwxr-xr-x 1        0 postgres  4096 Jan 27 14:12 ..
drwxrwxr-x 2        0 postgres  4096 Jan 27 14:12 conf.d
-rw-r--r-- 1 postgres postgres   207 Jan 27 15:14 pg_hba.conf
-rw-r--r-- 1 postgres postgres 26601 Jan 27 15:14 postgresql.conf
```

**Comment by blootsvoets** [comment]
To replicate from scratch:
sudo docker run -it --rm -e POSTGRESQL_PASSWORD=p -e POSTGRESQL_DATABASE=test -e POSTGRESQL_USERNAME=postgres -p "15432:5432" -v p_test:/bitnami/postgresql bitnami/postgresql:12.1.0

**Comment by blootsvoets** [comment]
Just as an update, the contents of the container has the following files and permissions. Group is set to `root`, user is set to `1001`.
```
sudo ls -anl local_p_test/data/
total 124
drwx------ 19 1001    0  4096 Jan 28 10:44 .
drwxr-xr-x  3 1001 1001  4096 Jan 28 10:44 ..
drwx------  5 1001    0  4096 Jan 28 10:44 base
drwx------  2 1001    0  4096 Jan 28 10:44 global
drwx------  2 1001    0  4096 Jan 28 10:44 pg_commit_ts
drwx------  2 1001    0  4096 Jan 28 10:44 pg_dynshmem
-rw-------  1 1001    0  4513 Jan 28 10:44 pg_hba.conf
-rw-------  1 1001    0  1636 Jan 28 10:44 pg_ident.conf
drwx------  4 1001    0  4096 Jan 28 10:44 pg_logical
drwx------  4 1001    0  4096 Jan 28 10:44 pg_multixact
drwx------  2 1001    0  4096 Jan 28 10:44 pg_notify
drwx------  2 1001    0  4096 Jan 28 10:44 pg_replslot
drwx------  2 1001    0  4096 Jan 28 10:44 pg_serial
drwx------  2 1001    0  4096 Jan 28 10:44 pg_snapshots
drwx------  2 1001    0  4096 Jan 28 10:44 pg_stat
drwx------  2 1001    0  4096 Jan 28 10:44 pg_stat_tmp
drwx------  2 1001    0  4096 Jan 28 10:44 pg_subtrans
drwx------  2 1001    0  4096 Jan 28 10:44 pg_tblspc
drwx------  2 1001    0  4096 Jan 28 10:44 pg_twophase
-rw-------  1 1001    0     3 Jan 28 10:44 PG_VERSION
drwx------  3 1001    0  4096 Jan 28 10:44 pg_wal
drwx------  2 1001    0  4096 Jan 28 10:44 pg_xact
-rw-------  1 1001    0    88 Jan 28 10:44 postgresql.auto.conf
-rw-------  1 1001    0 26623 Jan 28 10:44 postgresql.conf
```

**Comment by asfdtodd** [comment]
We're having this same issue in our k8s cluster. I can run 1.3.0 fine, but 1.3.1 and 1.4.0 fail with the same sed error.

**Comment by asfdtodd** [comment]
Sorry, let me be more clear; my issue is with postgresql-ha; but the base tag that was last working for me on this repository is 4.1.0-debian-9-r46, all of the debian-10 tags report this permission issue. I can run postgresql-ha 1.4.0 and set the tag for the bitnami/postgresql-repmgr repo to the 9-r46 tag and it works fine. So there's something wrong with the image itself.

**Comment by andresbono** [comment]
Hi @blootsvoets, @asfdtodd, we could finally reproduce your issue.

**TL;DR;** We released new versions of the container including a workaround. You can see it here: https://github.com/bitnami/bitnami-docker-postgresql/commit/4a93421c17b1b6181dbb9ea167fe778fb905295c. The chart was also updated: https://github.com/helm/charts/commit/1c06ec9e7e9ffc9596f4ea7ef83fe1192f8b67a9.

Could you please try these new versions and let us know if it works for you?

@asfdtodd, can you tell us what's the Linux distro, kernel version and Docker version where you are running the container? Thanks in advance.

**Comment by andresbono** [comment]
We noticed that the issue appears when sed with the "in-place" option is executed. To reproduce it:

```
docker run --rm --user 1001 bitnami/minideb:buster bash -c 'echo hi > /tmp/file && sed -i "s/hi/bye/" /tmp/file && cat /tmp/file'
cat: /tmp/file: Permission denied
```

For some reason, even with the appropriate permissions, the file can't be read after modifying it with sed. Other editing tools doesn't "break" the permissions over the file.

By re-setting the permissions (there is no need to change them to a new value), the file can be accessed again.

```
$ echo hi > /tmp/file && sed -i 's/hi/bye/' /tmp/file && cat /tmp/file
cat: /tmp/file: Permission denied
$ ls -l /tmp/file
-rw-r--r-- 1 1001 root 4 Jan 29 12:06 /tmp/file
$ chmod u+r /tmp/file
$ cat /tmp/file
bye
```

We tried by checking the system calls with strace or synchronizing the cache of the persistent storage but we couldn't determine the actual cause of the issue.

The issue only appears under specific combination of kernel/Docker/container/sed versions.

**Scenarios where the issue is reproducible**

  - Ubuntu `16.04` (Kernel `4.4.0-1098-aws`), Docker `18.09.7`.
    - Container `bitnami/minideb:buster`. Implies sed `4.7`.
    - Container `debian:10`. Implies sed `4.7`.

  - Ubuntu `16.04` (Kernel `4.4.0-1101-aws`), Docker `19.03.5`.
    - Container `bitnami/minideb:buster`. Implies sed `4.7`.
    - Container `debian:10`. Implies sed `4.7`.
    - Container `ubuntu:19.10`. Implies sed `4.7`.

  - Ubuntu `16.04.5`(Kernel `4.4.0-173`), Docker `19.03.5~3-0~ubuntu-xenial`.
    - Container `debian:10`. Implies sed `4.7`.

**Scenarios where the issue is NOT reproducible**

  - Ubuntu `16.04` (Kernel `4.4.0-148-generic`), Docker `18.03.0-ce`.
    - Container `bitnami/minideb:buster`. Implies sed `4.7`.
    - Container `debian:10`. Implies sed `4.7`.

  - Ubuntu `18.04` (Kernel `4.15.0-1058-aws`), Docker `19.03.5`.
    - Container `bitnami/minideb:buster`. Implies sed `4.7`.
    - Container `debian:10`. Implies sed `4.7`.
    - Container `ubuntu:19.10`. Implies sed `4.7`.

  - Ubuntu `19.10` (Kernel `5.3.0-29-generic`).
    -  Docker `19.03.3` (5:19.03.3~3-0~ubuntu-disco) + Container `debian:10` (sed `4.7`)
    -  Docker `18.06` (18.06.1~ce~3-0~ubuntu) + Container `debian:10` (sed `4.7`)

  - Ubuntu `16.04.5`(Kernel `4.4.0-131`), Docker `5:18.09.1~3-0~ubuntu-xenial`.
    - Container `debian:10`. Implies sed `4.7`.

**Comment by blootsvoets** [comment]
Thank you for the thorough investigation. This almost looks like a Docker bug rather than a bitnami one... Fixed in bitnami/postgresql:12.1.0 (`sha256:99f047e1b786ab2650baf7be19488bf4b00dcccd34e319877ca30c1af8179a4d`).

**Comment by andresbono** [comment]
Thanks for confirming it, @blootsvoets!
